### PR TITLE
Fix P2P not working for libraries

### PR DESCRIPTION
**Pull request by oscartbeaumont**
Changes:
 - [x] Completly rewrite `sd-p2p`
 - [x] Split everything up into more crates
 - [x] Decouple state
 - [x] Allow enabling and disabling mDNS independent to the rest of P2P
 - [x] Get core compiling
 - [x] Port libp2p over
 - [x] Fix connection logic -> Connection targets system
 - [x] Make changes to `P2P` emit a `HookEvent`
 - [x] Finish mDNS service removal
 - [x] Fix Spacedrop

TODO:
 - [x] Generate PCR migration
 - [x] Run Clippy
 - [x] Typescript

Breaking Changes:
 - This PR will wipe out the existing P2P keypair on the node due to the format of them changing. This shouldn't cause any issues to anyone unless they are using experimental feature flags.

### CODE REVIEW

File: core/src/lib.rs [hunk 1]
```diff
@@ -112,7 +112,9 @@ impl Node {
 		let (locations, locations_actor) = location::Locations::new();
 		let (jobs, jobs_actor) = job::Jobs::new();
 		let libraries = library::Libraries::new(data_dir.join("libraries")).await?;
-		let (p2p, p2p_actor) = p2p::P2PManager::new(config.clone(), libraries.clone()).await?;
+		let (p2p, start_p2p) = p2p::P2PManager::new(config.clone(), libraries.clone())
+			.await
+			.unwrap(); // TODO: Don't panic here
```
Comment (oscartbeaumont) [inline core/src/lib.rs:117]: This2

File: core/src/p2p/p2p_manager.rs [hunk 2]
```diff
@@ -1,169 +1,219 @@
...
-			sd_p2p::Manager::new(SPACEDRIVE_APP_ID, &keypair, manager_config).await?;
+		libraries::start(this.p2p.clone(), libraries);
 
 		info!(
 			"Node RemoteIdentity('{}') libp2p::PeerId('{}') is now online listening at addresses: {:?}",
-			manager.identity(),
-			manager.libp2p_peer_id(),
-			stream.listen_addrs()
+			this.p2p.remote_identity(),
+			"todo", // TODO: Work this out??? // TODO: Work out libp2p `PeerId`
```
Comment (oscartbeaumont) [inline core/src/p2p/p2p_manager.rs:61]: This3

File: core/src/p2p/p2p_manager.rs [hunk 3]
```diff
@@ -1,169 +1,219 @@
...
+			.find(|(i, p)| **i == identity && p.service().contains_key(&library_id))
+			.map(|(_, p)| p.clone())
 	}
 
-	pub async fn shutdown(&self) {
-		self.manager.shutdown().await;
+	pub fn state(&self) -> serde_json::Value {
+		json!({
+			"self_identity": self.p2p.remote_identity().to_string(),
+			// "self_peer_id": self.p2p.remote_identity().to_string(), // TODO
```
Comment (oscartbeaumont) [inline core/src/p2p/p2p_manager.rs:161]: This4

File: crates/p2p2/src/mdns.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+				discovered.insert(identity.clone(), peer);
+			}
+		}
+		ServiceEvent::ServiceRemoved(_, fullname) => {
+			let Some(identity) = fullname_to_identity(&state, &fullname) else {
+				return;
+			};
+
+			todo!();
```
Comment (oscartbeaumont) [inline crates/p2p2/src/mdns.rs:158]: Also this

File: crates/p2p2/src/peer.rs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	/// RemoteIdentity of the peer.
+	pub(crate) identity: RemoteIdentity,
+	/// Information from `P2P::service` on the remote node.
+	pub(crate) metadata: RwLock<HashMap<String, String>>,
+	/// We want these states to locked by the same lock so we can ensure they are consistent.
+	pub(crate) state: RwLock<State>,
+	/// A reference back to the P2P system.
+	/// This is weak so we don't have recursive `Arc`'s that can never be dropped.
+	pub(crate) p2p: Weak<P2P>,
+	// TODO: `pub removed: AtomicBool,` // TODO: This should disable methods on this `Peer` instance cause it can be cloned outta the system.
```
Comment (oscartbeaumont) [inline crates/p2p2/src/peer.rs:22]: Finish this

File: crates/p2p2/src/quic/transport.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		// We are assuming the libp2p `Keypair`` is the same format as our `Identity` type.
+		// This is *acktually* true but they reserve the right to change it at any point.
+		let keypair =
+			libp2p::identity::Keypair::ed25519_from_bytes(p2p.identity().to_bytes()).unwrap(); // TODO: Work out how to do this conversion
+		let libp2p_peer_id = Libp2pPeerId(keypair.public().to_peer_id());
+
+		let (tx, rx) = bounded(15);
+		let (internal_tx, internal_rx) = bounded(15);
+		let id = p2p.register_listener("libp2p-quic", tx, |peer, addrs| {
+			todo!();
```
Comment (oscartbeaumont) [inline crates/p2p2/src/quic/transport.rs:69]: dis

**Review by Brendonovich — APPROVED** [review]

**Pull request merged**